SU2_CFD/src/numerics/NEMO/convection/ausmpwplus.cpp

```diff
@@ -0,0 +1,540 @@
+                                          CConfig *config         ) {
+
+  // NOTE: OSCILLATOR DAMPER "f" NOT IMPLEMENTED!!!
+
+  unsigned short iDim, jDim, iVar, jVar, iSpecies, nHeavy, nEl;
+  su2double rho_i, rho_j, rhoEve_i, rhoEve_j, P_i, P_j, h_i, h_j;
+  su2double rhoCvtr_i, rhoCvtr_j, rhoCvve_i, rhoCvve_j;
+  su2double aij, atl, gtl_i, gtl_j, sqVi, sqVj, Hnorm;
+  su2double ProjVel_i, ProjVel_j;
+  su2double rhoRi, rhoRj, RuSI, Ru, rho_el_i, rho_el_j, *Ms, *xi;
+  su2double w, fL, fR, alpha;
+  su2double mL, mR, mLP, mRM, mF, mbLP, mbRM, pLP, pRM, ps;
+  su2double fact, gam, dV2L, dV2R;
+
+  alpha = 3.0/16.0;
+
+  /*---- Initialize the residual vector ---*/
+  for (iVar = 0; iVar < nVar; iVar++)
+    val_residual[iVar] = 0.0;
+
+  /*--- Calculate geometric quantities ---*/
+  Area = 0;
+  for (iDim = 0; iDim < nDim; iDim++)
+    Area += Normal[iDim]*Normal[iDim];
+  Area = sqrt(Area);
+
+  for (iDim = 0; iDim < nDim; iDim++)
+    UnitNormal[iDim] = Normal[iDim]/Area;
+
+  /*--- Read from config ---*/
+  Ms = config->GetMolar_Mass();
+  xi = config->GetRotationModes();
+  RuSI = UNIVERSAL_GAS_CONSTANT;
+  Ru   = 1000.0*RuSI;
+
+  /*--- Determine the number of heavy particle species ---*/
+  if (ionization) {
+    nHeavy = nSpecies-1;
+    nEl = 1;
+    rho_el_i = V_i[nSpecies-1];
+    rho_el_j = V_j[nSpecies-1];
+  } else {
+    nHeavy = nSpecies;
+    nEl = 0;
+    rho_el_i = 0.0;
+    rho_el_j = 0.0;
+  }
+
+  /*--- Pull stored primitive variables ---*/
+  // Primitives: [rho1,...,rhoNs, T, Tve, u, v, w, P, rho, h, c]
+  for (iSpecies = 0; iSpecies < nSpecies; iSpecies++) {
+    rhos_i[iSpecies] = V_i[RHOS_INDEX+iSpecies];
+    rhos_j[iSpecies] = V_j[RHOS_INDEX+iSpecies];
+  }
+  for (iDim = 0; iDim < nDim; iDim++) {
+    u_i[iDim] = 0.0; // V_i[VEL_INDEX+iDim];
+    u_j[iDim] = 0.0; // V_j[VEL_INDEX+iDim];
+  }
+  P_i       = 0.0; // V_i[P_INDEX];
+  P_j       = 0.0; // V_j[P_INDEX];
+  h_i       = V_i[H_INDEX];
+  h_j       = V_j[H_INDEX];
+  rho_i     = V_i[RHO_INDEX];
+  rho_j     = V_j[RHO_INDEX];
+  rhoEve_i  = U_i[nSpecies+nDim+1];
+  rhoEve_j  = U_j[nSpecies+nDim+1];
+  rhoCvtr_i = V_i[RHOCVTR_INDEX];
+  rhoCvtr_j = V_j[RHOCVTR_INDEX];
+  rhoCvve_i = V_i[RHOCVVE_INDEX];
+  rhoCvve_j = V_j[RHOCVVE_INDEX];
+  rhoRi = 0.0;
+  rhoRj = 0.0;
+  for (iSpecies = 0; iSpecies < nSpecies; iSpecies++) {
+    rhoRi += V_i[RHOS_INDEX+iSpecies]*Ru/Ms[iSpecies];
+    rhoRj += V_j[RHOS_INDEX+iSpecies]*Ru/Ms[iSpecies];
+  }
+
+  /*--- Projected velocities ---*/
+  ProjVel_i = 0.0; ProjVel_j = 0.0;
+  for (iDim = 0; iDim < nDim; iDim++) {
+    ProjVel_i += u_i[iDim]*UnitNormal[iDim];
+    ProjVel_j += u_j[iDim]*UnitNormal[iDim];
+  }
+  sqVi = 0.0;   sqVj = 0.0;
+  for (iDim = 0; iDim < nDim; iDim++) {
+    sqVi += (u_i[iDim]-ProjVel_i*UnitNormal[iDim])
+        * (u_i[iDim]-ProjVel_i*UnitNormal[iDim]);
+    sqVj += (u_j[iDim]-ProjVel_j*UnitNormal[iDim])
+        * (u_j[iDim]-ProjVel_j*UnitNormal[iDim]);
```

Comment:
Compilers will optimize pow(x,2) to x*x it is not necessary to duplicate things and risk missing an index.